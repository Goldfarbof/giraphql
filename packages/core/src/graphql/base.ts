import { GraphQLType } from 'graphql';
import BasePlugin from '../plugin';
import BuildCache from '../build-cache';

export default abstract class BaseType<
  Types extends GiraphQLSchemaTypes.TypeInfo,
  Name extends string | keyof Types['Input'] | keyof Types['Output'],
  Shape,
  InputShape = Shape,
  MatchShape = Shape
> {
  typename: Name;

  shape?: Shape;

  inputShape?: InputShape;

  matchShape?: MatchShape;

  abstract kind: 'Object' | 'Union' | 'Interface' | 'Enum' | 'Scalar' | 'InputObject';

  constructor(name: Name, shape?: Shape) {
    this.typename = name;
    this.shape = shape;
  }

  abstract buildType(cache: BuildCache<Types>, plugins: BasePlugin<Types>[]): GraphQLType;
}
