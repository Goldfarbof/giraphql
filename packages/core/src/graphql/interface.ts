import { GraphQLInterfaceType } from 'graphql';
import fromEntries from 'object.fromentries';
import BaseType from './base';
import { ShapeFromTypeParam, TypeParam, InterfaceName, ObjectName } from '../types';
import Field from '../field';
import FieldBuilder from '../fieldUtils/builder';
import ObjectType from './object';
import BasePlugin from '../plugin';
import BuildCache from '../build-cache';

export default class InterfaceType<
  Shape extends {},
  Types extends GiraphQLSchemaTypes.TypeInfo,
  Name extends InterfaceName<Types>
> extends BaseType<Types, Name, ShapeFromTypeParam<Types, Name, true>> {
  kind: 'Interface' = 'Interface';

  description?: string;

  options: GiraphQLSchemaTypes.InterfaceTypeOptions<{}, Types, any>;

  fieldShape?: Shape;

  constructor(name: Name, options: GiraphQLSchemaTypes.InterfaceTypeOptions<Shape, Types, Name>) {
    super(name);

    this.description = options.description;

    this.options = options;
  }

  getFields() {
    return this.options.shape(new FieldBuilder({}, this.typename));
  }

  buildType(cache: BuildCache<Types>, plugins: BasePlugin<Types>[]): GraphQLInterfaceType {
    let types: ObjectType<{}, any, Types, ObjectName<Types>>[];

    return new GraphQLInterfaceType({
      name: String(this.typename),
      description: this.description,
      resolveType: (obj: unknown) => {
        if (!types) {
          types = cache.getImplementers(this.typename);
        }

        for (const type of types) {
          if (type.isType(obj)) {
            return String(type.typename);
          }
        }

        return String(this.typename);
      },
      fields: () =>
        fromEntries(
          Object.entries(cache.getFields(this.typename)).map(([key, field]) => [
            key,
            (field as Field<{}, Types, TypeParam<Types>, TypeParam<Types>>).build(
              key,
              cache,
              plugins,
            ),
          ]),
        ),
    });
  }
}
