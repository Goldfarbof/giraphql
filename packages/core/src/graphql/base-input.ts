import BaseType from './base';

export default abstract class BaseInputType<Shape = unknown, InputShape = Shape> extends BaseType<
  Shape
> {
  shape?: Shape;

  inputShape?: InputShape;

  abstract kind: 'Enum' | 'Scalar' | 'InputObject';
}
