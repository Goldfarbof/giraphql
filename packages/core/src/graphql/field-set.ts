import { FieldsShape } from '../types';

export default class FieldSet<Types extends GiraphQLSchemaTypes.TypeInfo, Shape = unknown> {
  kind: 'FieldSet' = 'FieldSet';

  forType: string;

  shape: FieldsShape<any, any>;

  constructor(type: string, shape: FieldsShape<Types, Shape>) {
    this.forType = type;
    this.shape = shape;
  }
}
