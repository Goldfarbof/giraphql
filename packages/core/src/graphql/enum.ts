import { GraphQLEnumType, GraphQLEnumValueConfigMap, GraphQLEnumValueConfig } from 'graphql';
// @ts-ignore
import fromEntries from 'object.fromentries';
import BaseType from './base';
import { EnumValues, ShapeFromEnumValues } from '../types';

export default class EnumType<
  Types extends GiraphQLSchemaTypes.TypeInfo,
  Name extends string,
  Values extends EnumValues = EnumValues,
  Shape extends string = ShapeFromEnumValues<Values>
> extends BaseType<Types, Name, Shape> {
  values: GraphQLEnumValueConfigMap;

  description?: string;

  kind: 'Enum' = 'Enum';

  options: GiraphQLSchemaTypes.EnumTypeOptions<EnumValues>;

  constructor(name: Name, options: GiraphQLSchemaTypes.EnumTypeOptions<Values>) {
    super(name);

    this.options = options;

    this.values = Array.isArray(options.values)
      ? fromEntries(options.values.map(key => [key, {}]))
      : fromEntries(
          Object.entries(options.values)
            .map(([key, value]) => {
              if (value && typeof value === 'object') {
                return [key, value];
              }

              if (typeof value === 'string') {
                return [value, {}];
              }

              return null;
            })
            .filter(Boolean) as [string, GraphQLEnumValueConfig][],
        );

    this.description = options.description;
  }

  buildType() {
    return new GraphQLEnumType({
      description: this.description,
      name: this.typename,
      values: this.values,
    });
  }
}
