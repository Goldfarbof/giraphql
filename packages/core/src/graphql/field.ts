/* eslint-disable @typescript-eslint/consistent-type-assertions */
import { GraphQLFieldConfig, GraphQLFieldConfigArgumentMap } from 'graphql';
// @ts-ignore
import fromEntries from 'object.fromentries';
import {
  TypeParam,
  InputFields,
  ShapeFromTypeParam,
  FieldNullability,
  FieldOptionsFromKind,
} from '../types';
import { typeFromParam, buildArg } from '../utils';
import BaseType from './base';
import BasePlugin from '../plugin';
import { BuildCache } from '..';

export default class Field<
  Args extends InputFields<Types>,
  Types extends GiraphQLSchemaTypes.TypeInfo,
  ParentType extends TypeParam<Types>,
  Type extends TypeParam<Types>,
  Nullable extends FieldNullability<Types, Type> = FieldNullability<Types, Type>,
  Extends extends string | null = null,
  Kind extends 'Object' | 'Interface' | 'Root' | 'Subscription' =
    | 'Object'
    | 'Interface'
    | 'Root'
    | 'Subscription',
  Options extends FieldOptionsFromKind<
    Types,
    ParentType,
    Type,
    Nullable,
    Args,
    Kind
  > = FieldOptionsFromKind<Types, ParentType, Type, Nullable, Args, Kind>
> {
  shape?: ShapeFromTypeParam<Types, Type, true>;

  nullable: Nullable;

  args: Args = {} as Args;

  extendsField: Extends;

  type: Type;

  options: Options;

  parentTypename: string;

  constructor(
    options: Options & {
      extendsField?: Extends;
    },
    parentTypename: string,
  ) {
    this.options = options;
    this.nullable = options.nullable || (false as Nullable);
    this.args = options.args ? options.args! : ({} as Args);
    this.extendsField = options.extendsField || (null as Extends);
    this.type = options.type;
    this.parentTypename = parentTypename;
  }

  protected buildArgs(cache: BuildCache<Types>): GraphQLFieldConfigArgumentMap {
    return fromEntries(
      Object.keys(this.args).map(key => {
        const arg = this.args[key];

        return [
          key,
          {
            description:
              typeof arg !== 'object' || arg instanceof BaseType || Array.isArray(arg)
                ? undefined
                : arg.description,
            required:
              typeof arg !== 'object' || arg instanceof BaseType || Array.isArray(arg)
                ? false
                : arg.required || false,
            type: buildArg(arg, cache),
            defaultValue:
              typeof arg !== 'object' || arg instanceof BaseType || Array.isArray(arg)
                ? undefined
                : arg.default,
          },
        ];
      }),
    );
  }

  build(
    name: string,
    cache: BuildCache<Types>,
    plugins: BasePlugin<Types>[],
  ): GraphQLFieldConfig<unknown, unknown> {
    const baseConfig: GraphQLFieldConfig<unknown, unknown> = {
      args: this.buildArgs(cache),
      description: this.options.description,
      resolve:
        cache.resolverMock(this.parentTypename, name) ||
        (this.options as { resolve?: (...args: unknown[]) => unknown }).resolve ||
        (() => {
          throw new Error(`Not implemented: No resolver found for ${this.parentTypename}.${name}`);
        }),
      subscribe:
        cache.subscribeMock(this.parentTypename, name) ||
        (this.options as { subscribe?: (...args: unknown[]) => unknown }).subscribe,
      type: typeFromParam(this.type, cache, this.nullable),
      extensions: this.options.extensions,
    };

    return plugins.reduce((config, plugin) => {
      return plugin.updateFieldConfig
        ? plugin.updateFieldConfig(name, this as any, config, cache)
        : config;
    }, baseConfig);
  }
}
