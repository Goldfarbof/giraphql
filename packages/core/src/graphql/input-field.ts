import { GraphQLInputFieldConfig, GraphQLArgumentConfig } from 'graphql';
// @ts-ignore
import { inputTypeFromParam } from '../utils';
import { BuildCache, InputType, SchemaTypes, partialInputShapeKey, PartialInput } from '..';

export default class InputField<T> implements PartialInput {
  kind: 'InputField' = 'InputField';

  [partialInputShapeKey]: T;

  required: boolean | { list: boolean; items: boolean };

  type: InputType<SchemaTypes> | [InputType<SchemaTypes>];

  options: GiraphQLSchemaTypes.InputOptions;

  constructor(options: GiraphQLSchemaTypes.InputOptions) {
    this.options = options;
    this.required = options.required ?? false;
    this.type = options.type;
  }

  // This is currently used to for fields on Input Objects, and Arguments on Fields
  build(cache: BuildCache): GraphQLInputFieldConfig & GraphQLArgumentConfig {
    return {
      description: this.options.description,
      defaultValue: this.options.defaultValue,
      extensions: this.options.extensions,
      type: inputTypeFromParam(this.type, cache, this.required),
    };
  }
}
