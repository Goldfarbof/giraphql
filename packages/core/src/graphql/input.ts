import { GraphQLInputObjectType, GraphQLInputFieldConfigMap } from 'graphql';
// @ts-ignore
import fromEntries from 'object.fromentries';
import { InputFields } from '../types';
import BaseType from './base';
import { buildArg } from '../utils';
import InputFieldBuilder from '../fieldUtils/input';
import BuildCache from '../build-cache';

export default class InputObjectType<
  Types extends GiraphQLSchemaTypes.TypeInfo,
  Shape,
  Fields extends InputFields<Types>,
  Name extends string,
  MatchShape = Shape
> extends BaseType<Types, Name, Shape, Shape, MatchShape> {
  kind: 'InputObject' = 'InputObject';

  fields: Fields;

  options: GiraphQLSchemaTypes.InputTypeOptions<Types, InputFields<Types>>;

  constructor(name: Name, options: GiraphQLSchemaTypes.InputTypeOptions<Types, Fields>) {
    super(name);

    this.fields = options.shape(new InputFieldBuilder());

    this.options = options;
  }

  buildFields(cache: BuildCache<Types>): GraphQLInputFieldConfigMap {
    return fromEntries(
      Object.keys(this.fields).map(key => {
        const field = this.fields[key];
        return [
          key,
          {
            description:
              typeof field !== 'object' || field instanceof BaseType || Array.isArray(field)
                ? undefined
                : field.description,
            required:
              typeof field !== 'object' || field instanceof BaseType || Array.isArray(field)
                ? false
                : field.required || false,
            type: buildArg(field, cache),
          },
        ];
      }),
    );
  }

  buildType(cache: BuildCache<Types>) {
    return new GraphQLInputObjectType({
      name: this.typename,
      description: this.options.description,
      fields: () => this.buildFields(cache),
      extensions: this.options.extensions,
    });
  }
}
