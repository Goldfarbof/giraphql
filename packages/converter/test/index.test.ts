import exampleSchema from '@giraphql/core/test/examples/random-stuff';
import starwarsSchema from '@giraphql/core/test/examples/starwars/schema';
import { spawn } from 'child_process';
import { printSchema } from 'graphql';
import GirphQLConverter from '../src';

function execTS(script: string) {
  return new Promise<string>((resolve, reject) => {
    const chunks: Buffer[] = [];
    const child = spawn('ts-node', {
      stdio: ['pipe', 'pipe', 'inherit'],
    });

    child.on('error', err => reject(err));
    child.stdout.on('data', chunk => chunks.push(chunk));
    child.stdout.on('end', () => resolve(Buffer.concat(chunks).toString()));

    child.stdin.write(script);
    child.stdin.end();
  });
}

async function printGeneratedSchema(converter: GirphQLConverter) {
  const script = `import { printSchema } from 'graphql'\n${converter.toString()}\n\nconsole.log(printSchema(schema))`;

  const result = await execTS(script);

  return result;
}

describe('Code generator', () => {
  test('example schema', async () => {
    const converter = new GirphQLConverter(exampleSchema);

    expect(converter.toString()).toMatchSnapshot();

    const result = await printGeneratedSchema(converter);

    expect(result).toMatchSnapshot();
    expect(result.trim()).toEqual(printSchema(exampleSchema).trim());
  }, 10000);

  test('starwars schema', async () => {
    const converter = new GirphQLConverter(starwarsSchema);

    expect(converter.toString()).toMatchSnapshot();

    const result = await printGeneratedSchema(converter);

    expect(result).toMatchSnapshot();
    expect(result.trim()).toEqual(printSchema(starwarsSchema).trim());
  }, 10000);
});
