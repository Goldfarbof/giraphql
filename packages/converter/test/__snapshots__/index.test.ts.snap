// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code generator example schema 1`] = `
"import SchemaBuilder from \\"@giraphql/core\\";
const builder = new SchemaBuilder<{
    Context: {};
    Object: {
        User: unknown;
        Article: unknown;
        Sheep: unknown;
    };
    Interface: {
        Shaveable: unknown;
        Countable: unknown;
    };
    Input: {
        Example2: Example2Shape;
    };
}>();
const SearchResult = builder.createUnionType('SearchResult', {
    members: ['User', 'Article'],
    resolveType: (parent, context, info) => {
        throw new Error('Not implemented');
    }
});
const stuff = builder.createEnumType('stuff', {
    values: {
        Beats: {
            value: 'Beats',
        },
        Bears: {
            value: 'Bears',
        },
        BattlestarGalactica: {
            value: 'BattlestarGalactica',
        },
    },
});
const Example = builder.createInputType('Example', {
    shape: t => ({
        id: t.type('ID', {
            required: true,
        }),
        id2: t.type('Int', {
            required: false,
        }),
        ids: t.type(['ID'], {
            required: { list: true, items: true },
        }),
        ids2: t.type(['Int'], {
            required: { list: false, items: true },
        }),
    }),
});
interface Example2Shape {
    example: {
        id: string;
        id2: number | null | undefined;
        ids: string[];
        ids2: number[] | null | undefined;
    };
    id: string | null | undefined;
    ids: string[];
    more: Example2Shape;
}
const Example2 = builder.createInputType('Example2', {
    shape: t => ({
        example: t.type(Example, {
            required: true,
        }),
        id: t.type('ID', {
            required: false,
        }),
        ids: t.type(['ID'], {
            required: { list: true, items: true },
        }),
        more: t.type('Example2', {
            required: true,
        }),
    }),
});
const Shaveable = builder.createInterfaceType('Shaveable', {
    shape: t => ({
        id: t.field({
            type: 'ID',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        shaved: t.field({
            type: 'Boolean',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        extendMePlease: t.field({
            type: 'String',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        })
    }),
});
const Countable = builder.createInterfaceType('Countable', {
    shape: t => ({
        count: t.field({
            args: {
                max: t.arg.type('Int', {
                    required: true,
                }),
            },
            type: 'Int',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        })
    }),
});
const Query = builder.createQueryType({
    shape: t => ({
        user: t.field({
            type: 'User',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        stuff: t.field({
            type: [stuff],
            nullable: { list: false, items: false },
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        sheep: t.field({
            type: 'Sheep',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        })
    }),
});
const User = builder.createObjectType('User', {
    isType: (obj, context, info) => { throw new Error('Not implemented'); },
    shape: t => ({
        id: t.field({
            type: 'ID',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        displayName: t.field({
            type: 'String',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        firstName: t.field({
            type: 'String',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        lastName: t.field({
            type: 'String',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        firstBornChild: t.field({
            type: 'User',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        partialName: t.field({
            args: {
                example: t.arg.type(Example, {
                    required: true,
                }),
                firstN: t.arg.type('Int', {
                    required: true,
                }),
            },
            type: 'String',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        recursiveArgs: t.field({
            args: {
                example2: t.arg.type('Example2', {
                    required: true,
                }),
                firstN: t.arg.type('ID', {
                    required: false,
                }),
            },
            type: 'ID',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        related: t.field({
            type: SearchResult,
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        friends: t.field({
            type: ['User'],
            nullable: { list: false, items: false },
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        stuff: t.field({
            type: ['String'],
            nullable: { list: false, items: false },
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        optional: t.field({
            type: 'String',
            nullable: true,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        list: t.field({
            args: {
                ids: t.arg.type(['ID'], {
                    required: { list: true, items: true },
                }),
            },
            type: ['ID'],
            nullable: { list: false, items: false },
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        sparseList: t.field({
            args: {
                ids: t.arg.type(['ID'], {
                    required: { list: true, items: false },
                }),
            },
            type: ['ID'],
            nullable: { list: false, items: true },
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        notSparseList: t.field({
            args: {
                ids: t.arg.type(['ID'], {
                    required: { list: true, items: true },
                }),
            },
            type: ['ID'],
            nullable: { list: true, items: true },
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        })
    }),
});
const Article = builder.createObjectType('Article', {
    isType: (obj, context, info) => { throw new Error('Not implemented'); },
    shape: t => ({
        title: t.field({
            description: \\"Title of the article, probably click bait\\",
            type: 'String',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        })
    }),
});
const Sheep = builder.createObjectType('Sheep', {
    implements: [Shaveable, Countable],
    isType: (obj, context, info) => { throw new Error('Not implemented'); },
    shape: t => ({
        color: t.field({
            args: {
                id: t.arg.type('ID', {
                    required: false,
                }),
            },
            type: 'String',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        thing: t.field({
            type: stuff,
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        })
    }),
});
const Subscription = builder.createSubscriptionType({
    shape: t => ({
        event: t.field({
            type: 'String',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
            subscribe: (parent, args, context, info) => { throw new Error('Not implemented'); },
        })
    }),
});
export const schema = builder.toSchema([
    Query,
    User,
    Example,
    Example2,
    SearchResult,
    Article,
    stuff,
    Sheep,
    Shaveable,
    Countable,
    Subscription,
]);
"
`;

exports[`Code generator example schema 2`] = `
"type Article {
  \\"\\"\\"Title of the article, probably click bait\\"\\"\\"
  title: String!
}

interface Countable {
  count(max: Int!): Int!
}

input Example {
  id: ID!
  id2: Int
  ids: [ID!]!
  ids2: [Int!]
}

input Example2 {
  example: Example!
  id: ID
  ids: [ID!]!
  more: Example2!
}

type Query {
  user: User!
  stuff: [stuff!]!
  sheep: Sheep!
}

union SearchResult = User | Article

interface Shaveable {
  id: ID!
  shaved: Boolean!
  extendMePlease: String!
}

type Sheep implements Shaveable & Countable {
  id: ID!
  shaved: Boolean!
  extendMePlease: String!
  count(max: Int!): Int!
  color(id: ID): String!
  thing: stuff!
}

enum stuff {
  Beats
  Bears
  BattlestarGalactica
}

type Subscription {
  event: String!
}

type User {
  id: ID!
  displayName: String!
  firstName: String!
  lastName: String!
  firstBornChild: User!
  partialName(example: Example!, firstN: Int!): String!
  recursiveArgs(example2: Example2!, firstN: ID): ID!
  related: SearchResult!
  friends: [User!]!
  stuff: [String!]!
  optional: String
  list(ids: [ID!]!): [ID!]!
  sparseList(ids: [ID]!): [ID]!
  notSparseList(ids: [ID!]!): [ID]
}

"
`;

exports[`Code generator starwars schema 1`] = `
"import SchemaBuilder from \\"@giraphql/core\\";
const builder = new SchemaBuilder<{
    Context: {};
    Object: {
        Human: unknown;
        Droid: unknown;
    };
    Interface: {
        Character: unknown;
    };
}>();
const Episode = builder.createEnumType('Episode', {
    description: \\"One of the films in the Star Wars Trilogy\\",
    values: {
        NEWHOPE: {
            description: \\"Released in 1977.\\",
            value: 4,
        },
        EMPIRE: {
            description: \\"Released in 1980.\\",
            value: 5,
        },
        JEDI: {
            description: \\"Released in 1983\\",
            value: 6,
        },
    },
});
const MoreEpisode = builder.createEnumType('MoreEpisode', {
    values: {
        NEWHOPE: {
            description: \\"Released in 1977.\\",
            value: 4,
        },
        EMPIRE: {
            description: \\"Released in 1980.\\",
            value: 5,
        },
        JEDI: {
            description: \\"Released in 1983\\",
            value: 6,
        },
        OTHER: {
            value: 'OTHER',
        },
    },
});
const Character = builder.createInterfaceType('Character', {
    shape: t => ({
        id: t.field({
            description: \\"The id of the character\\",
            type: 'ID',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        name: t.field({
            description: \\"The name of the character\\",
            type: 'String',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        friends: t.field({
            description: \\"The friends of the character, or an empty list if they have none.\\",
            type: ['Character'],
            nullable: { list: false, items: false },
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        appearsIn: t.field({
            description: \\"Which movies they appear in.\\",
            args: {
                id: t.arg.type('ID', {
                    required: true,
                }),
            },
            type: [Episode],
            nullable: { list: false, items: false },
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        })
    }),
});
const Query = builder.createQueryType({
    shape: t => ({
        hero: t.field({
            args: {
                episode: t.arg.type(Episode, {
                    description: \\"If omitted, returns the hero of the whole saga. If provided, returns the hero of that particular episode.\\",
                    required: true,
                }),
            },
            type: 'Character',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        human: t.field({
            args: {
                id: t.arg.type('ID', {
                    description: \\"id of the character\\",
                    required: true,
                }),
            },
            type: 'Human',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        }),
        droid: t.field({
            args: {
                id: t.arg.type('ID', {
                    description: \\"id of the character\\",
                    required: true,
                }),
            },
            type: 'Droid',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        })
    }),
});
const Human = builder.createObjectType('Human', {
    description: \\"A humanoid creature in the Star Wars universe.\\",
    implements: [Character],
    isType: (obj, context, info) => { throw new Error('Not implemented'); },
    shape: t => ({
        homePlanet: t.field({
            description: \\"The home planet of the human, or null if unknown.\\",
            type: 'String',
            nullable: true,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        })
    }),
});
const Droid = builder.createObjectType('Droid', {
    description: \\"A mechanical creature in the Star Wars universe.\\",
    implements: [Character],
    isType: (obj, context, info) => { throw new Error('Not implemented'); },
    shape: t => ({
        primaryFunction: t.field({
            description: \\"The primary function of the droid.\\",
            type: 'String',
            nullable: false,
            resolve: (parent, args, context, info) => { throw new Error('Not implemented'); },
        })
    }),
});
export const schema = builder.toSchema([
    Query,
    Episode,
    Character,
    Human,
    Droid,
    MoreEpisode,
]);
"
`;

exports[`Code generator starwars schema 2`] = `
"interface Character {
  \\"\\"\\"The id of the character\\"\\"\\"
  id: ID!

  \\"\\"\\"The name of the character\\"\\"\\"
  name: String!

  \\"\\"\\"The friends of the character, or an empty list if they have none.\\"\\"\\"
  friends: [Character!]!

  \\"\\"\\"Which movies they appear in.\\"\\"\\"
  appearsIn(id: ID!): [Episode!]!
}

\\"\\"\\"A mechanical creature in the Star Wars universe.\\"\\"\\"
type Droid implements Character {
  \\"\\"\\"The id of the character\\"\\"\\"
  id: ID!

  \\"\\"\\"The name of the character\\"\\"\\"
  name: String!

  \\"\\"\\"The friends of the character, or an empty list if they have none.\\"\\"\\"
  friends: [Character!]!

  \\"\\"\\"Which movies they appear in.\\"\\"\\"
  appearsIn(id: ID!): [Episode!]!

  \\"\\"\\"The primary function of the droid.\\"\\"\\"
  primaryFunction: String!
}

\\"\\"\\"One of the films in the Star Wars Trilogy\\"\\"\\"
enum Episode {
  \\"\\"\\"Released in 1977.\\"\\"\\"
  NEWHOPE

  \\"\\"\\"Released in 1980.\\"\\"\\"
  EMPIRE

  \\"\\"\\"Released in 1983\\"\\"\\"
  JEDI
}

\\"\\"\\"A humanoid creature in the Star Wars universe.\\"\\"\\"
type Human implements Character {
  \\"\\"\\"The id of the character\\"\\"\\"
  id: ID!

  \\"\\"\\"The name of the character\\"\\"\\"
  name: String!

  \\"\\"\\"The friends of the character, or an empty list if they have none.\\"\\"\\"
  friends: [Character!]!

  \\"\\"\\"Which movies they appear in.\\"\\"\\"
  appearsIn(id: ID!): [Episode!]!

  \\"\\"\\"The home planet of the human, or null if unknown.\\"\\"\\"
  homePlanet: String
}

enum MoreEpisode {
  \\"\\"\\"Released in 1977.\\"\\"\\"
  NEWHOPE

  \\"\\"\\"Released in 1980.\\"\\"\\"
  EMPIRE

  \\"\\"\\"Released in 1983\\"\\"\\"
  JEDI
  OTHER
}

type Query {
  hero(
    \\"\\"\\"
    If omitted, returns the hero of the whole saga. If provided, returns the hero of that particular episode.
    \\"\\"\\"
    episode: Episode!
  ): Character!
  human(
    \\"\\"\\"id of the character\\"\\"\\"
    id: ID!
  ): Human!
  droid(
    \\"\\"\\"id of the character\\"\\"\\"
    id: ID!
  ): Droid!
}

"
`;
